### DISCUSSION TITLE
Hook on update

### DISCUSSION THREAD

**Discussion by razbakov** [prisma/prisma#1890] [Q&A] [answered]
Is it possible to add backend functions in prisma on mutations like it is in graphcool?

**Comment by nikolasburk** [comment, marked as answer]
Hey @razbakov, _functions_ are not supported in Prisma 2.0. That's mostly because Prisma 2.0 is a _very_ different product than Graphcool, so there's actually no need for this feature. 

If you still think functions would be useful in Prisma 2.0, can you provide an example use case? Also, if you want to read up on what Prisma 2.0 is and how it works, I recommend checking out the [new docs](https://prisma2.netlify.com/) (will be published soon).

**Reply by razbakov** [reply]
Hey @nikolasburk  thank you for reply!

Use cases would be:
- trigger mailgun to send an email once table row created, changed or deleted (user sign up, rsvp to event, etc)
- add async rabbitmq job to trigger other service to do some job (and to establish communication between different microservices)
- update some system fields only in backend (example: accepted payment increases account.balance, which is read-only in UI)

Is possible to do without functions, how?

**Reply by nikolasburk** [reply]
These are not really use cases for Prisma 2.0 though since Prisma 2.0 is "just" a database client for Node.js and TypeScript (similar to an ORM), so you would implement all of these features manually in your application server.

Unlike Graphcool, Prisma 2.0 is no backend-as-a-service. You need to implement your own API layer, Prisma 2.0 only helps you connect it to your database. Does that make sense? 🙂

**Reply by razbakov** [reply]
Do you have some example code + demo of app like this?

**Reply by nikolasburk** [reply]
You can find lots of examples [here](https://github.com/prisma/prisma-examples/) 👍

**Reply by nikolasburk** [reply]
I definitely recommend you check out the [new docs](https://prisma2.netlify.com/) 🙂

**Reply by razbakov** [reply]
Thank you!

I think I will be not the only one to ask this question.
It might be useful to add section `hooks` or `functions` in the docs and highlight your answer from here.
Because migrating from old mindset of graphcool to new prisma is a bit confusing.

**Comment by ChristianRich** [comment]
For a hobby project, I'm building a mini e-commerce system built with Prisma.
I have models such as Customer, Cart, Product, Payment, LineItem etc . You get the idea.

Here are some of my top use cases for Prisma hooks:
- Re-calculate cart total amount and tax amount, when Cart model is updated
- Trigger user activation email when a new User is created
- Trigger a password reset email, when a new PasswordRequest model is created
- Trigger an email to admin when a SystemEvent is created and the status equals ALERT

Hooks are great for decoupling things, giving it a more "event-driven" feel.

I imagine something like this:

```
prisma.user.onUpdate = async (user: User | null ) => {
  if (user && user.status === 'NEW_SIGNUP') {
    await sendActivationEmail(user);
  }
}
```

Btw, thanks for building an amazing product and outstanding documentation!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
